src/gprs.c

```diff
@@ -536,9 +542,228 @@ static void pri_context_signal_settings(struct pri_context *ctx,
 				context_settings_append_ipv6);
 }
 
-static void pri_parse_proxy(struct pri_context *ctx, const char *proxy)
+static void set_route(const struct context_settings *settings,
+							const char *ipstr)
+{
+	struct rtentry rt;
+	struct sockaddr_in addr;
+	int sk;
+
+	/* TODO Handle IPv6 case */
+
+	DBG("%s", ipstr);
+
+	if (settings->interface == NULL)
+		return;
+
+	sk = socket(PF_INET, SOCK_DGRAM, 0);
+	if (sk < 0)
+		return;
+
+	memset(&rt, 0, sizeof(rt));
+	rt.rt_flags = RTF_UP | RTF_HOST;
+	rt.rt_dev = (char *) settings->interface;
+
+	memset(&addr, 0, sizeof(addr));
+	addr.sin_family = AF_INET;
+	addr.sin_addr.s_addr = inet_addr(ipstr);
+
+	if (addr.sin_addr.s_addr == INADDR_NONE) {
+		ofono_error("Cannot set route for invalid IP %s", ipstr);
+		return;
+	}
+
+	memcpy(&rt.rt_dst, &addr, sizeof(rt.rt_dst));
+
+	memset(&addr, 0, sizeof(addr));
+	addr.sin_family = AF_INET;
+	addr.sin_addr.s_addr = INADDR_ANY;
+	memcpy(&rt.rt_gateway, &addr, sizeof(rt.rt_gateway));
+
+	memset(&addr, 0, sizeof(addr));
+	addr.sin_family = AF_INET;
+	addr.sin_addr.s_addr = INADDR_ANY;
+	memcpy(&rt.rt_genmask, &addr, sizeof(rt.rt_genmask));
+
```

Comment:
I was going to suggest why you didn't just set addr.sin_family and addr.sin_addr.s_addr once for gateway and netmask, and then I noticed that you did this in delete_route.

I'll point out that most of the code in these two functions is duplicated, so perhaps it'd be better to collapse into a single function that handles adding and removing routes based on a third parameter?

Also, you probably don't need to specify the RTF_UP flag when deleting a route.